src/business-logic/services/device-connection-service-implementation.ts

```diff
@@ -0,0 +1,100 @@
+
+    await deviceConnection.connect()
+
+    this.connectedDevices.set(device.id, {deviceConnection, device})
+
+    return DeviceConnectionStatus.CONNECTED
+  }
+
+  public async send(_deviceId: string, _params: string[]): Promise<void> {
+    const conDevice: DeviceAggregate | undefined = this.connectedDevices.get(_deviceId)
+    if(conDevice === undefined) throw new DeviceNotFoundError(_deviceId)
+
+    await conDevice?.deviceConnection.send(_deviceId, _params)
+  }
+
+  public async listen(_deviceId: string, _callback: (data: unknown) => void): Promise<void> {
+    const conDevice: DeviceAggregate | undefined = this.connectedDevices.get(_deviceId)
+    if(conDevice === undefined) throw new DeviceNotFoundError(_deviceId)
+
+    await conDevice?.deviceConnection.listen(_deviceId, _callback)
+  }
+
+  public getConnectionStatusById(deviceId: string): DeviceConnectionStatus {
+    // we don't do thorough testing of both actual connection and the isConnected field since isConnected is an inheritance from CoreDevice and not needed here.
+    return this.connectedDevices.has(deviceId) ? DeviceConnectionStatus.CONNECTED : DeviceConnectionStatus.DISCONNECTED 
+  }
+
+  public async disconnectConnectionById(deviceId: string): Promise<DeviceConnectionStatus> {
+    await this.connectedDevices.get(deviceId)?.deviceConnection.disconnect()
+    return DeviceConnectionStatus.DISCONNECTED
+  }
+
+  public async removeConnectionById(deviceId: string): Promise<DeviceConnectionStatus> {
+    const device: DeviceAggregate | undefined = this.connectedDevices.get(deviceId)
+    if (!device?.deviceConnection) throw new DeviceAlreadyRemovedError(deviceId)
+
+    await device.deviceConnection.disconnect()
+
+    this.connectedDevices.delete(deviceId)
+
+    return DeviceConnectionStatus.DISCONNECTED
+  }
+
+  public connectionExists(deviceId: string): boolean {
+    return this.connectedDevices.has(deviceId)
+  }
+
+  public listAllNetworkedDevices(): Device[] {
+    const networkedDevices: Device[] = []
+    this.connectedDevices.forEach((deviceAggregate, _key) => {
+      networkedDevices.push(deviceAggregate.device)
+    })
+    return networkedDevices
+  }
+}
+
+class DeviceAlreadyConnectedError extends Error {
+  constructor(deviceId: string) {
+    super(`Device with ID '${deviceId}' is already connected.`)
+    this.name = 'DeviceAlreadyConnectedError'
+  }
+}
+
+class DeviceAlreadyRemovedError extends Error {
+  constructor(deviceId: string) {
+    super(`Device with ID '${deviceId}' is already removed.`)
+    this.name = 'DeviceAlreadyRemovedError'
+  }
+}
+
+class DeviceNotFoundError extends Error {
+  constructor(deviceId: string) {
+    super(`Device with ID '${deviceId}' is not in the collection. Have you forgot to create the connection?`)
+    this.name = 'DeviceAlreadyRemovedError'
+  }
+}
+
+export enum DeviceConnectionStatus {
+  DISCONNECTED = 0,
+  CONNECTED = 1
+}
```

Comment:
We have a convention for internal enums that they the value is the same as the key. So in this case it would be
```typescript
export enum DeviceConnectionStatus {
  DISCONNECTED = 'DISCONNECTED',
  CONNECTED = 'CONNECTED',
}
```

Both this and the DeviceXXXErrors should be moved to separate files, since they are part of defining the contract with the abstraction and not specific for this implementation :)

Reply:
Wow, that was unexpected. Does they still convert to truthy and falsy values? What is the benefit of doing it this way? (value same as key)

I suppose magic numbers and readability will be mentioned, but going against the entire IT industry enum standard, the way every programmer since birth has been taught to use enums, does require some argumentation, methinks. 

Will comply with the code standard, but I am struggling to make sense of this.

Reply:
The primary reason for going with this is that it removes a layer of abstraction.
E.g. If I save a device configuration with the attribute `type: 22`, then the probability of the observer knowing what that means is quite low. If the attribute instead says `INEWS_GATEWAY`, then everyone – even non-programmers knows what kind of device we are talking about.
It does have the drawback of having a larger memory footprint, but that is also a consequence of writing out attribute names that we do as well.

Do they need to have boolean representations?
Let us say that a device has the attribute `connectionStatus`. In the case of the condition `device.connectionStatus`, would it check the existence of a connection status value or the state of said value? Another aspect is that a connection can be in more states than `CONNECTED` and `DISCONNECTED`. In the case where the configuration is missing, it could be `UNKNOWN`/`UNDEFINED`; or we could choose to model the transition states as well `CONNECTING`/`DISCONNECTING`. In these cases, what would a boolean representation mean?

I haven't talked with every programmer nor learned every curriculum for courses and talks about enums, and I haven't been in the industry for that long. So you might have a point 😁 But a lot of programming languages have quite good support for non-numerical enumerations and personally I was not indoctrinated with numerical enumerations at my education.
And as you mention, they can help against magic literals and with improving readability. If it is something you feel strongly about, then lets have a chat in team about it 😃

Reply:
I was imprecise and said boolean when I meant numerical. Sloppy of me, I know.

Not sure if enums are the way to go if we want to divorce ourselves from numerical representations, but that's another topic.

I will follow your suggested standard.